### PR TITLE
Allow two-way bindings to signal-based template variables

### LINKED ISSUE #54670: two-way binding with a signal in @for-block throws a compilation error
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

Yes

### Description

Two-way binding with a signal in @for-block throws a compilation error. Below is an example. In this example, `nameModels` is a Signal of the Signal array. Imagine a use case like FormArray for reactive forms. The template binds an input element to each element of `nameModels`. This example works perfectly when written with isolated bindings. However, replacing it with the "banana-in-a-box" syntax results in a compile error. (Stackblitz will not finish loading waiting for compilation)

> Error: Illegal state: unsupported expression in two-way action binding.

```ts
@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule, FormsModule],
  template: `
    <h1>two-way binding with signals in for-block</h1>
    <div>
      @for(item of nameModels(); track $index){
        <!-- compile error -->
        <!-- <input type="text" [(ngModel)]="item" /> -->
        
        <!-- works -->
        <input
          type="text"
          [ngModel]="item()"
          (ngModelChange)="item.set($event)"
        />
      }
    </div>
  `,
})
export class App {
  nameModels: WritableSignal<WritableSignal<string>[]> = signal([
    signal('Angular'),
  ]);
}
```

This use case may not be common, but it is conceivable enough, and I believe the compiler should support this use case.
Since Signal is an object, not a primitive, temporary variables in iterations with for blocks should be pass-by-reference and modifiable. In fact, you can explicitly write `item.set()`.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-mpactd?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
✘ [ERROR] Angular compilation emit failed. [plugin angular-compiler]

  Error: Illegal state: unsupported expression in two-way action binding.
      at wrapAssignmentAction (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:7087:11)
      at convertAssignmentActionBinding (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:7024:23)
      at prepareEventListenerParameters (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:28169:9)
      at Object.paramsOrFn (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:29795:20)
      at getInstructionStatements (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:5195:84)
      at TemplateDefinitionBuilder.buildTemplateFunction (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:28372:36)
      at file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:28908:50
      at file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:28350:60
      at Array.forEach (<anonymous>)
      at TemplateDefinitionBuilder.buildTemplateFunction (file:///Users/lacolaco/works/ng172signals/node_modules/@angular/compiler/fesm2022/compiler.mjs:28350:33)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 17.2.2
Node: 20.11.0
Package Manager: npm 10.2.4
OS: darwin arm64

Angular: 17.2.3
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1702.2
@angular-devkit/build-angular   17.2.2
@angular-devkit/core            17.2.2
@angular-devkit/schematics      17.2.2
@angular/cli                    17.2.2
@schematics/angular             17.2.2
rxjs                            7.8.1
typescript                      5.3.3
zone.js                         0.14.4
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by crisbeto**
Includes the following changes whose goal is to properly support two-way bindings to signal-based local variables:

### refactor(compiler-cli): introduce template semantics checker
Introduces a new `TemplateSemanticsChecker` that will be used to flag semantic errors in the user's template. Currently we do some of this in the type check block, but the problem is that it doesn't have access to the template type checker which prevents us from properly checking cases like https://github.com/angular/angular/issues/54670. This pass is also distinct from the extended template checks, because we don't want users to be able to turn the checks off and we want them to run even if `strictTemplates` are disabled.

### refactor(compiler-cli): move signal identification function
Moves the function that identifies signals into a separate file so that it can be reused outside of extended diagnostics.

### refactor(compiler-cli): move illegal template assignment check into template semantics checker
Moves the check which ensures that there are no writes to template variables into the `TemplateSemanticsChecker` to prepare for the upcoming changes.

### fix(compiler-cli): flag two-way bindings to non-signal values in templates
We have a diagnostic that reports writes to template variables which worked both for regular event bindings and two-way bindings, however the latter was broken by https://github.com/angular/angular/pull/54154 because two-way bindings no longer had a `PropertyWrite` AST.

These changes fix the diagnostic and expand it to allow two-way bindings to template variables that are signals.

### fix(compiler): handle two-way bindings to signal-based template variables in instruction generation
Updates the instruction generation for two-way bindings to only emit the `twoWayBindingSet` call when writing to template variables. Since template variables are constants, it's only allowed to write to them when they're signals. Non-signal values are flagged during template type checking.

Fixes https://github.com/angular/angular/issues/54670.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts [hunk 1]
```diff
@@ -534,15 +534,8 @@ export class TraitCompiler implements ProgramTypeCheckAdapter {
...
-        continue;
-      }
-      const record = this.classes.get(clazz)!;
-      for (const trait of record.traits) {
-        if (trait.handler.extendedTemplateCheck === undefined) {
-          continue;
-        }
-        diagnostics.push(...trait.handler.extendedTemplateCheck(clazz, extendedTemplateChecker));
+      if (isNamedClassDeclaration(clazz)) {
+        diagnostics.push(...checker.getDiagnosticsForComponent(clazz));
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:538]: Won't this run for all classes that have a trait attached to them? i.e. might be a pipe for example

File: packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts [hunk 2]
```diff
@@ -213,6 +213,12 @@ export interface TemplateTypeChecker {
    */
   invalidateClass(clazz: ts.ClassDeclaration): void;
 
+  /**
+   * Gets the target of a template expression, if possible.
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/typecheck/api/checker.ts:217]: Would love this doc to link to `BoundTarget.getExpressionTarget` — this only handles local refs or template variables

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+              .name}' as the left-hand side of an assignment expression. Template variables are read-only.`;
+      this.diagnostics.push(this.makeIllegalTemplateVarDiagnostic(target, context, errorMessage));
+    }
+  }
+
+  private checkForIllegalWriteInTwoWayBinding(ast: PropertyRead, context: TmplAstNode) {
+    // Only check top-level property reads inside two-way bindings for illegal assignments.
+    if (!(context instanceof TmplAstBoundEvent) || context.type !== ParsedEventType.TwoWay ||
+        !(ast.receiver instanceof ImplicitReceiver) ||
+        ast !== unwrapAstWithSource(context.handler)) {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:87]: why is this needed? Do you know in what cases it's wrapped?

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    const target = this.templateTypeChecker.getExpressionTarget(ast, this.component);
+    if (!(target instanceof TmplAstVariable)) {
+      return;
+    }
+
+    // Two-way bindings to template variables are only allowed if the variables are signals.
+    const symbol = this.templateTypeChecker.getSymbolOfNode(target, this.component);
+    if (symbol !== null && !isSignalReference(symbol)) {
```
Comment (devversion) [inline packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:98]: Should we always error if `symbol === null` and we don't know, except that it's a template variable?

File: packages/compiler/src/compiler_util/expression_converter.ts [hunk 5]
```diff
@@ -84,14 +84,24 @@ export function convertAssignmentActionBinding(
...
-      .or(ast.set(EventHandlerVars.event));
+function wrapAssignmentReadExpression(ast: o.ReadPropExpr|o.ReadKeyExpr|
+                                      o.ReadVarExpr): o.Expression {
+  const setOperation =
+      new o.ExternalExpr(R3.twoWayBindingSet).callFn([ast, EventHandlerVars.event]);
+
+  // If the expression is targeting a variable read, we only emit the `twoWayBindingSet` since
+  // the fallback would be attempting to write into a constant. Invalid usages will be flagged
+  // during template type checking.
+  if (ast instanceof o.ReadVarExpr) {
```
Comment (devversion) [inline packages/compiler/src/compiler_util/expression_converter.ts:95]: Is `ReadVarExpr` always a constant? I see it as a variable, not a constant? This seems like a large assumption from the `resolve_names` phase?

`o.variable()` creates such AST object

**Review by devversion — APPROVED** [review]
LGTM overall, but a couple of comments

File: packages/core/test/acceptance/authoring/model_inputs_spec.ts [hunk 6]
```diff
@@ -589,4 +589,44 @@ describe('model inputs', () => {
     fixture.detectChanges();
     expect(() => fixture.destroy()).not.toThrow();
   });
+
+  it('should support two-way binding to a signal @for loop variable', () => {
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/authoring/model_inputs_spec.ts:593]: I would _love_ to see a test illustrating how it behaves with non-signal items in a collection.

This is also an interesting consideration of how do we see items produced by a loop - are those assignable or not (signals are, obviously)

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 3 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:87]: It's a bit of a gotcha since `ASTWithSource extends AST`. I don't know exactly when we use one over the other, but there are several places where we need to do this sort of unwrapping.

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 4 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:98]: I was a bit concerned that there may be cases where we don't have full type information and we would raise the error unnecessarily.

File: packages/core/test/acceptance/authoring/model_inputs_spec.ts (hunk 6 above)
Comment (crisbeto) [reply packages/core/test/acceptance/authoring/model_inputs_spec.ts:593]: So a non-signal value would basically be a no-op at runtime and we have a compiler diagnostic to flag it at compile time.

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (crisbeto) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: It can be either a `let` or `const`. Either way we need to not re-assign it (which is what this change is doing), but just pass it to the `twoWayBindingSet` instruction.

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 4 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:98]: Interesting. maybe worth trying in g3?

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (devversion) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: But isn't this a huge assumption? A variable is technically assignable right?

We are just assuming it to be a constant template variable because the compiler converts the LexicalReadExpr into a ReadVarExpr

I might miss something

Comment (crisbeto) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: Template variables have always been read-only, we even have a diagnostic for it. That's why I think it's safe to assume that it would be a constant here as well.

File: packages/compiler-cli/src/ngtsc/transform/src/compilation.ts (hunk 1 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/transform/src/compilation.ts]: Good point. I've reworked it.

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 4 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts]: Sure, will run a TGP now.

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (devversion) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: Oh yeah, I think that is true. I guess my concern is that we are assuming that `ReadVarExpr` indicates a template variable?

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 3 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:87]: I see!

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (crisbeto) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: Even if it's not a template variable, I think we'd want to cover it with this check. E.g. I believe that `$event` would also be a `ReadVarExpr` which shouldn't be writable either.

File: packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts (hunk 4 above)
Comment (crisbeto) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:98]: Okay so even without the TGP it appears to be problematic. It was flagging one of our own `TestBed`-based tests, because it can't resolve the symbol there.

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/typecheck/template_semantics/src/template_semantics_checker.ts:98]: Sounds good

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (devversion) [reply packages/compiler/src/compiler_util/expression_converter.ts:95]: Right, but `ReadVarExpr` could technically also be some arbitrary `var` variable (I know we don't generate this right now— but it _could_ be)

**Comment by crisbeto** [comment]
[Passing TGP](http://OCL:613817804:BASE:613885343:1709899217435:319b3de7)

File: packages/compiler/src/compiler_util/expression_converter.ts (hunk 5 above)
Comment (crisbeto) [reply packages/compiler/src/compiler_util/expression_converter.ts:97]: Based on our discussion over DM: I think that this is a safe assumption to make at the moment. I've added added a comment stating that it's an assumption. Will continue as is for now.

**Review by devversion — APPROVED** [review]

**Comment by atscott** [comment]
This PR was merged into the repository by commit 5ae2bf480697b475c41ea136f08f4dca633c75b9.
